clang/runtime/dpct-rt/include/dpl_extras/vector.h

```diff
@@ -66,143 +67,191 @@ class device_vector {
 
   void _set_capacity_and_alloc() {
     _capacity = ::std::max(_size * 2, _min_capacity());
-    _storage = _alloc.allocate(_capacity);
+    _storage = ::std::allocator_traits<Allocator>::allocate(_alloc, _capacity);
+  }
+
+  void _construct(size_type n, size_type start_idx = 0) {
+    if (n > 0) {
+      for (size_type i = 0; i < n; i++) {
+        ::std::allocator_traits<Allocator>::construct(_alloc,
+                                                      _storage + start_idx + i);
+      }
+    }
+  }
+
+  void _construct(size_type n, const T &value, size_type start_idx = 0) {
+    if (n > 0) {
+      for (size_type i = 0; i < n; i++) {
+        ::std::allocator_traits<Allocator>::construct(
+            _alloc, _storage + start_idx + i, value);
+      }
+    }
+  }
+
+  template <typename Iter>
+  void _construct(Iter first, Iter last, size_type start_idx = 0) {
+    size_type num_eles = ::std::distance(first, last);
+    if (num_eles > 0) {
+      size_type i = 0;
+      for (auto iter = first; iter != last; iter++, i++) {
+        ::std::allocator_traits<Allocator>::construct(
+            _alloc, _storage + start_idx + i, *iter);
+      }
+    }
+  }
+
+  void _destroy(size_type n, size_type start_idx = 0) {
+    if (n > 0) {
+      for (size_type i = 0; i < n; i++) {
+        ::std::allocator_traits<Allocator>::destroy(_alloc,
+                                                    _storage + start_idx + i);
+      }
+    }
   }
 
 public:
   template <typename OtherA> operator ::std::vector<T, OtherA>() const {
     auto __tmp = ::std::vector<T, OtherA>(this->size());
     ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
```

Comment:
Here the objects construction is speeded up by dpl algorithm, but you comments said that is not make sense. So which is expected?

Reply:
This instance is copying elements, not constructing them.  It does not have the issues described below.

The issue with construction being sped up by a SYCL kernel or oneDPL algorithm is two-fold: 
1) Conceptually, allocators (which may be custom user-supplied overrides of `usm_allocator`) can't be assumed to have thread safe `construct` calls.  For instance, if you had a constructor which counted the number of element constructions and did not provide synchronization around the count variable, this could have issues.
2) Technically, `std::allocator_traits<T>::construct()` is a variadic function, and those cannot be used SYCL device functions.  Therefore, it cannot be called within a SYCL kernel.  As I mention before, if parallel accellerated construction is a feature which is highly requested, we can create our own version of `allocator_traits` which does not rely on variadic functions for use in this context.  However, with (1) in mind, it does not seem worthwhile to add this complexity.  

If we add this, we would need to state some rules about what is allowed in custom `construct` calls.